### PR TITLE
Fix IO.fx eager evaluation

### PR DESCRIPTION
**Pull request by ps-feng**
The previous implementation was launching a coroutine straight away and running the provided code until the first suspension point.

This was causing the issue described in https://github.com/arrow-kt/arrow/issues/1617, which can be summarized as
```kotlin
val someIO = IO.fx {
    someFun() // <-- called on caller's thread because it's evaluated straight away
}

IO.fx {
    continueOn(Dispatchers.IO)
    someIO.bind()
}.unsafeRunAsync { }
```

### CODE REVIEW

**Comment by pakoito** [comment]
Could you please see if there are other `fx` functions with the same issue, maybe in the Rx or Reactor packages?

**Comment by ps-feng** [comment]
The issue is indeed quite extended, I'll see if I can fix them all in this PR.

**Comment by ps-feng** [comment]
This is trickier than I expected. Actually the initial problem I reported isn't solved because most of the monads' `fx` extension use `MonadFx.monad`, which launches the coroutine immediately.
```kotlin
interface MonadFx<F> {
  val M: Monad<F>
  fun <A> monad(c: suspend MonadSyntax<F>.() -> A): Kind<F, A> {
    val continuation = MonadContinuation<F, A>(M)
    val wrapReturn: suspend MonadContinuation<F, *>.() -> Kind<F, A> = { just(c()) }
    wrapReturn.startCoroutine(continuation, continuation)
    return continuation.returnedMonad()
  }
}
```
That's fine for most monads, but not so much for `IO` or transformers that contain `IO`, so my original case is still not fixed by the change suggested in this PR because it's using precisely `MonadFx.monad`:
```kotlin
EitherT.monad(IO.monad()).fx.monad {
    whatever() // <-- eagerly evaluated, then suspends
    someEitherT.bind()
}
```
Any ideas?

**Comment by pakoito** [comment]
`fx.monad` is correct because those don't allow suspension. The ones that do are supposed to be overriden with that version.

File: modules/fx/arrow-fx/src/test/kotlin/arrow/fx/IOTest.kt [hunk 1]
```diff
@@ -282,6 +282,17 @@ class IOTest : UnitSpec() {
         .unsafeRunSync()
     }
 
+    "fx should defer evaluation until run" {
```
Comment (JorgeCastilloPrz) [inline modules/fx/arrow-fx/src/test/kotlin/arrow/fx/IOTest.kt]: 👍

**Comment by JorgeCastilloPrz** [comment]
Shouldn't an `EitherT` for IO allow suspension @pakoito? I was under the impression that it should given it's an `IO` empowered with `Either` capabilities, but still an `IO`.

**Comment by pakoito** [comment]
Ooooooh, that's a tricky one. @raulraja do you know how it is in Scala?

**Comment by raulraja** [comment]
Nine of this matter faster October when the comprehension plugin is out since binding and suspension will be unrelated. Binding it's gonna be translated to flatMap and the only way to inject suspension at compile time is via effect or the IO constructor which does not eagerly run. All the fx interfaces and typeclass will be gone since fx is gonna be powered by synthetic codegen via compiler plugins. As for the current fix whatever works I'm fine with but we can't change signatures now to change them back in October. It would force us into a deprecation cycle we don't want to through before meta and compiler plugins take over the FP style

**Comment by ps-feng** [comment]
If that's the case, should we just close this ticket and wait for the plugin?

**Comment by JorgeCastilloPrz** [comment]
The fix doesn't seem to change signatures, but deprecation is needed for breaking changes not just on signatures. Since when we've got this bug @nomisRev? If it's been there since long ago the fix could also be considered a breaking change given it drastically changes the method behaviour. In the other hand, If it's been there since 0.10.0 only then I say we fix it fast since it seems to be a quite important bug from my point of view.

**Comment by pakoito** [comment]
Let's merge with this. We should fix Rx and Reactor too, if they have the same problem.

**Comment by ps-feng** [comment]
I'm not familiar with coroutines but I believe they have the same problem.
```kotlin
fun <A> MonoK.Companion.fx(c: suspend AsyncSyntax<ForMonoK>.() -> A): MonoK<A> =
  MonoK.async().fx.async(c).fix()

interface AsyncFx<F> : MonadThrowFx<F> {
  val async: Async<F>
  override val ME: MonadThrow<F> get() = async
  fun <A> async(c: suspend AsyncSyntax<F>.() -> A): Kind<F, A> {
    val continuation = AsyncContinuation<F, A>(async)
    val wrapReturn: suspend AsyncSyntax<F>.() -> Kind<F, A> = { just(c()) }
    wrapReturn.startCoroutine(continuation, continuation) // <-- will execute until next bind I think
    return continuation.returnedMonad()
  }
}
```

**Comment by pakoito** [comment]
Yep, so just wrap in `defer` for now.

**Comment by pakoito** [comment]
Actually, wait, it's

fun <A> async(c: suspend AsyncSyntax<F>.() -> A): Kind<F, A>

what should be wrapped.

**Comment by ps-feng** [comment]
I think it makes no difference but I'll check tomorrow.

**Comment by ps-feng** [comment]
I did verify that the same issue was occuring in Rx and Reactor, so I created a separate PR for them: https://github.com/arrow-kt/arrow/pull/1663

**Comment by pakoito** [comment]
We should have merged this already, damn. I'll check the build.

**Pull request merged**